### DISCUSSION TITLE
sharpe returns based style analysis

### DISCUSSION THREAD

**Discussion by Jacques2101** [skfolio/skfolio#83] [Q&A] [answered]
Hi i would like to know if it is possible to estimate Return based style analysis with skfolio ie : 
Find (w1,w2,...wn) such as min tracking-error of (Rp - sum(wi*Fi)) with wi>=0 and sum(wi)=1

**Comment by HugoDelatte** [comment, marked as answer]
Hello,

It’s possible to do this indirectly by using the `MeanRisk` estimator, depending on how you define such tracking error.

This problem belongs to a more general class of problems focused on finding an asset proxy. In your case, the asset you are trying to proxy is your portfolio (in our example, we will use the S&P 500 Index), and the proxy universe consists of some factors.

Finding the proxy weights can be viewed as a linear regression with additional constraints on the weights. But it can also be seen as a portfolio optimization problem, where the factors plus your portfolio are optimized with the portfolio's weight fixed to -1 (that's the residuals from your regression)

So, you simply need to minimize the RMSE of this new portfolio. However, the RMSE risk measure is not implemented, so you will need to minimize the variance. If the mean of the residuals is 0, it will give you the same results. Otherwise, it’s "almost" similar to performing RMSE minimization with or without an intercept.

I'll add RMSE to the risk measures because it can be useful for this type of proxy analysis.

Now, the advantage of transforming the initial problem into a portfolio optimization problem is that you can leverage all the existing features, such as imposing additional constraints or minimizing different risk measures that focus more on tail risks, such as CVaR (it gives you another view on the factor exposure of your portfolio which can be quite interesting).

Here’s an example:
 
 ```python
 from skfolio import RiskMeasure
from skfolio.datasets import load_sp500_index, load_factors_dataset
from skfolio.preprocessing import prices_to_returns
from skfolio.optimization import MeanRisk, ObjectiveFunction

factor_prices = load_factors_dataset()
sp500_prices = load_sp500_index()

X, y = prices_to_returns(factor_prices, sp500_prices)

Xy = X.join(y)

model = MeanRisk(
    objective_function=ObjectiveFunction.MINIMIZE_RISK,
    risk_measure=RiskMeasure.VARIANCE,
    budget=0,
    min_weights=-1,
    max_weights=1,
    groups={
        "MTUM": "factors",
        "QUAL": "factors",
        "SIZE": "factors",
        "USMV": "factors",
        "VLUE": "factors",
    },
    linear_constraints=["SP500 <= -1", "factors >= 0"],
)
model.fit(Xy)
print(model.weights_)

 ````

**Comment by Jacques2101** [comment]
Thanks but i get an error message which is :

AttributeError: 'csc_matrix' object has no attribute 'A'


{
      "name": "AttributeError",
      "message": "'csc_matrix' object has no attribute 'A'",
      "stack": "---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], line 20
      3 Xy = X.join(y)
      5 model = MeanRisk(
      6    objective_function=ObjectiveFunction.MINIMIZE_RISK,
      7    risk_measure=RiskMeasure.VARIANCE,
   (...)
     18    linear_constraints=[\"SP500 <= -1\", \"factors >= 0\"],
     19 )
---> 20 model.fit(Xy)
     21 print(model.weights_)

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\skfolio\\optimization\\convex\\_mean_risk.py:975, in MeanRisk.fit(self, X, y)
    972 problem = cp.Problem(objective, constraints)
    974 # results
--> 975 self._solve_problem(
    976     problem=problem,
    977     w=w,
    978     factor=factor,
    979     parameters_values=parameters_values,
    980     expressions={
    981         \"expected_return\": expected_return,
    982         \"risk\": risk,
    983         \"mu_uncertainty_set\": mu_uncertainty_set,
    984         \"regularization\": regularization,
    985         \"factor\": factor,
    986     },
    987 )
    989 return self

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\skfolio\\optimization\\convex\\_base.py:951, in ConvexOptimization._solve_problem(self, problem, w, factor, parameters_values, expressions)
    949 with warnings.catch_warnings():
    950     warnings.simplefilter(\"ignore\")
--> 951     problem.solve(solver=self.solver, **self._solver_params)
    953 if w.value is None:
    954     raise cp.SolverError(\"No solution found\")

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\problems\\problem.py:503, in Problem.solve(self, *args, **kwargs)
    501 else:
    502     solve_func = Problem._solve
--> 503 return solve_func(self, *args, **kwargs)

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\problems\\problem.py:1072, in Problem._solve(self, solver, warm_start, verbose, gp, qcp, requires_grad, enforce_dpp, ignore_dpp, canon_backend, **kwargs)
   1069         self.unpack(chain.retrieve(soln))
   1070         return self.value
-> 1072 data, solving_chain, inverse_data = self.get_problem_data(
   1073     solver, gp, enforce_dpp, ignore_dpp, verbose, canon_backend, kwargs
   1074 )
   1076 if verbose:
   1077     print(_NUM_SOLVER_STR)

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\problems\\problem.py:696, in Problem.get_problem_data(self, solver, gp, enforce_dpp, ignore_dpp, verbose, canon_backend, solver_opts)
    693     s.LOGGER.info(
    694              'Compiling problem (target solver=%s).', solver_name)
    695     s.LOGGER.info('Reduction chain: %s', reduction_chain_str)
--> 696 data, inverse_data = solving_chain.apply(self, verbose)
    697 safe_to_cache = (
    698     isinstance(data, dict)
    699     and s.PARAM_PROB in data
    700     and not any(isinstance(reduction, EvalParams)
    701                 for reduction in solving_chain.reductions)
    702 )
    703 self._compilation_time = time.time() - start

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\reductions\\chain.py:76, in Chain.apply(self, problem, verbose)
     74     if verbose:
     75         s.LOGGER.info('Applying reduction %s', type(r).__name__)
---> 76     problem, inv = r.apply(problem)
     77     inverse_data.append(inv)
     78 return problem, inverse_data

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\reductions\\dcp2cone\\cone_matrix_stuffing.py:335, in ConeMatrixStuffing.apply(self, problem)
    333 # Form the constraints
    334 extractor = CoeffExtractor(inverse_data, self.canon_backend)
--> 335 params_to_P, params_to_c, flattened_variable = self.stuffed_objective(
    336     problem, extractor)
    337 # Lower equality and inequality to Zero and NonNeg.
    338 cons = []

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\reductions\\dcp2cone\\cone_matrix_stuffing.py:322, in ConeMatrixStuffing.stuffed_objective(self, problem, extractor)
    319 if self.quad_obj:
    320     # extract to 0.5 * x.T * P * x + q.T * x + r
    321     expr = problem.objective.expr.copy()
--> 322     params_to_P, params_to_c = extractor.quad_form(expr)
    323     # Handle 0.5 factor.
    324     params_to_P = 2*params_to_P

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\utilities\\coeff_extractor.py:235, in CoeffExtractor.quad_form(self, expr)
    231 quad_forms = replace_quad_forms(root, {})
    233 # Calculate affine parts and combine them with quadratic forms to get
    234 # the coefficients.
--> 235 coeffs, constant = self.extract_quadratic_coeffs(root.args[0],
    236                                                  quad_forms)
    237 # Restore expression.
    238 restore_quad_forms(root.args[0], quad_forms)

File c:\\Users\\JacquesTebeka\\miniforge3\\envs\\data\\lib\\site-packages\\cvxpy\\utilities\\coeff_extractor.py:137, in CoeffExtractor.extract_quadratic_coeffs(self, affine_expr, quad_forms)
    135 constant = param_coeffs[-1, :]
    136 # TODO keep sparse.
--> 137 c = param_coeffs[:-1, :].A
    138 num_params = param_coeffs.shape[1]
    140 # coeffs stores the P and q for each quad_form,
    141 # as well as for true variable nodes in the objective.

AttributeError: 'csc_matrix' object has no attribute 'A'"
}

Cordialement


Jacques Tebeka

 Tél. : +33 6 28 69 19 01


________________________________
De : Hugo Delatte ***@***.***>
Envoyé : mardi 24 septembre 2024 11:22
À : skfolio/skfolio ***@***.***>
Cc : Jacques ***@***.***>; Author ***@***.***>
Objet : Re: [skfolio/skfolio] sharpe returns based style analysis (Discussion #83)


Hello,

It’s possible to do this indirectly by using the MeanRisk estimator, depending on how you define such tracking error.

This problem belongs to a more general class of problems focused on finding an asset proxy. In your case, the asset you are trying to proxy is your portfolio (in our example, we will use the S&P 500 Index), and the proxy universe consists of some factors.

Finding the proxy weights can be viewed as a linear regression with additional constraints on the weights. But it can also be seen as a portfolio optimization problem, where the factors plus your portfolio are optimized with the portfolio's weight fixed to -1 (that's the residuals from your regression)

So, you simply need to minimize the RMSE of this new portfolio. However, the RMSE risk measure is not implemented, so you will need to minimize the variance. If the mean of the residuals is 0, it will give you the same results. Otherwise, it’s similar to performing RMSE minimization with or without an intercept.

I'll add RMSE to the risk measures because it can be useful for this type of proxy analysis.

Now, the advantage of transforming the initial problem into a portfolio optimization problem is that you can leverage all the existing features, such as imposing additional constraints or minimizing different risk measures that focus more on tail risks, such as CVaR (it gives you another view on the factor exposure of your portfolio which can be quite interesting).

Here’s an example:

from skfolio import RiskMeasure
from skfolio.datasets import load_sp500_index, load_factors_dataset
from skfolio.preprocessing import prices_to_returns
from skfolio.optimization import MeanRisk, ObjectiveFunction

factor_prices = load_factors_dataset()
sp500_prices = load_sp500_index()

X, y = prices_to_returns(factor_prices, sp500_prices)

Xy = X.join(y)

model = MeanRisk(
   objective_function=ObjectiveFunction.MINIMIZE_RISK,
   risk_measure=RiskMeasure.VARIANCE,
   budget=0,
   min_weights=-1,
   max_weights=1,
   groups={
       "MTUM": "factors",
       "QUAL": "factors",
       "SIZE": "factors",
       "USMV": "factors",
       "VLUE": "factors",
   },
   linear_constraints=["SP500 <= -1", "factors >= 0"],
)
model.fit(Xy)
print(model.weights_)

**Comment by HugoDelatte** [comment]
Seems to be coming from an old version of scipy or cvxpy. Could you confirm your version of the following dependencies, or test by updating them to their latest versions:

numpy
scipy
pandas
cvxpy
scikit-learn
joblib
plotly

**Comment by Jacques2101** [comment]
after some upgrade it works

Thx !

Cordialement


Jacques Tebeka

 Tél. : +33 6 28 69 19 01


________________________________
De : Hugo Delatte ***@***.***>
Envoyé : mardi 24 septembre 2024 11:51
À : skfolio/skfolio ***@***.***>
Cc : Jacques ***@***.***>; Author ***@***.***>
Objet : Re: [skfolio/skfolio] sharpe returns based style analysis (Discussion #83)


Seems to be coming from an old version of scipy. Could you confirm your version of the following dependencies, or test by updating them to their latest versions:

numpy
scipy
pandas
cvxpy
scikit-learn
joblib
plotly
